### PR TITLE
Llama3-70B LoRA multi GPU

### PR DESCRIPTION
**Pull request by rohan-varma**
#### Context
- Adds LoRA support for distributed training of Llama3-70B model.
- Note that HF currently bundles both the meta and HF checkpoints into a single repo for llama3. This makes downloading only a single model a little tricky (and we should avoid user having to download 2 copies of model). For llama3-8B, the documented download commands work out of the box since safetensors are by default ignored by `tune download` and we use the meta checkpointer.
- However, using the meta checkpointer for 70B model is much more complicated and requires significant development work to get it to support sharded checkpoints and unsharding/resharding. We can get 70B support much quicker if we use the HF checkpoints and safe tensors. Hence, we're documenting (currently in configs only)how to run the download command to get the safetensors and NOT the meta checkpoint files (so users only have 1 copy of the model), and using the meta checkpointer as the default.

#### Changelog
- Render recipe in recipe registry
- Add config for 70B LoRA training
- NOTE: need to download with `--ignore-patterns "original/consolidated*"` to be able to get the safetensors
- Update README and docs as appropriate

#### Test plan
- ```tune download meta-llama/Meta-Llama-3-70b --hf-token <> --output-dir /tmp/Meta-Llama-3-70b --ignore-patterns "original/consolidated*"```
- ```tune run --nproc_per_node 8 lora_finetune_distributed --config recipes/configs/llama3/70B_lora.yaml```
- Loss curve: 
image
- Can't easily do eval due to lack of distributed support. Generation result is pending.
- Render in `tune ls` - 
image
- Some logs - 

```
2024-04-18:13:49:43,502 INFO     [lora_finetune_distributed.py:280] FSDP is enabled. Instantiating Model on CPU for Rank 0 ...
2024-04-18:13:51:59,203 INFO     [lora_finetune_distributed.py:286] Model instantiation took 135.70 secs
2024-04-18:13:53:10,922 INFO     [lora_finetune_distributed.py:360] Memory Stats after model init:
{'peak_memory_active': 28.329404416, 'peak_memory_alloc': 26.228041728, 'peak_memory_reserved': 32.377929728}
```
- Peak memory allocated is < 30 GB on GPU 0:
image

- Peak memory reserved on GPU 0:
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,7 +6,7 @@
 &nbsp;
 &nbsp;
 
-torchtune now officially supports Meta Llama3! Check out our recipes for Llama3-8B with LoRA, QLoRA and Full fine-tune in the [Llama3](#llama3) section! 🚀 🦙
+torchtune now officially supports Meta Llama3! Check out our recipes for Llama3-8B with LoRA (70B also supported), QLoRA and Full fine-tune in the [Llama3](#llama3) section! 🚀 🦙
```
Comment (ebsmothers) [inline README.md:9]: Nit: would actually just put 70B in a separate sentence, like "We also support 70B fine-tuning with LoRA"

File: README.md [hunk 2]
```diff
@@ -86,35 +86,41 @@ This table captures the minimum memory requirements for our different recipes us
 
 ## Llama3
 
-torchtune supports fine-tuning for the Llama3 8B models with support for 70B on its way. We currently support LoRA, QLoRA and Full-finetune on a single GPU as well as LoRA and Full fine-tune on multiple devices. For all the details, take a look at our [tutorial](https://pytorch.org/torchtune/main/tutorials/llama3.html).
+torchtune supports fine-tuning for the Llama3 8B and 70B models. We currently support LoRA, QLoRA and Full-finetune on a single GPU as well as LoRA and Full fine-tune on multiple devices for the 8B model, and LoRA on multiple devices for the 70B model. For all the details, take a look at our [tutorial](https://pytorch.org/torchtune/main/tutorials/llama3.html).
 
 
-In our initial experiments, QLoRA has a peak allocated memory of ``~9GB`` while LoRA on a single GPU has a peak allocated memory of ``~19GB``. To get started, you can use our default configs to kick off training.
+In our initial experiments for Llama3 8B, QLoRA has a peak allocated memory of ``~9GB`` while LoRA on a single GPU has a peak allocated memory of ``~19GB``. To get started, you can use our default configs to kick off training.
```
Comment (ebsmothers) [inline README.md:92]: nit
```suggestion
In our initial experiments for Llama3-8B, QLoRA has a peak allocated memory of ``~9GB`` while LoRA on a single GPU has a peak allocated memory of ``~19GB``. To get started, you can use our default configs to kick off training.
```

File: README.md [hunk 3]
```diff
@@ -86,35 +86,41 @@ This table captures the minimum memory requirements for our different recipes us
...
 ```
 
-- Full fine-tune on 2 GPUs
+- 8B Full fine-tune on 2 GPUs
 
 ```bash
 tune run --nproc_per_node 2 full_finetune_distributed --config llama3/8B_full
 ```
 
+- 70B LoRA finetune on 8 GPUs
```
Comment (ebsmothers) [inline README.md:118]: Not sure what the right place to do this is, but do we wanna mention memory requirements to run 70B somewhere?

File: recipes/configs/llama3/70B_lora.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#   tune download meta-llama/Meta-Llama-3-70b --hf-token <TOKEN> --output-dir /tmp/Meta-Llama-3-70b --ignore-patterns "original/consolidated*"
+#
+# This config needs 8 GPUs to run
+#   # tune run --nproc_per_node 8 lora_finetune_distributed --config recipes/configs/llama3/70B_lora.yaml
+#
+
+# Model Arguments
+model:
+  _component_: torchtune.models.llama3.lora_llama3_70b
+  lora_attn_modules: ['q_proj', 'v_proj', 'k_proj']
```
Comment (ebsmothers) [inline recipes/configs/llama3/70B_lora.yaml:15]: nit: I feel like this is clearer
```suggestion
  lora_attn_modules: ['q_proj', 'k_proj', 'v_proj']
```

File: recipes/configs/llama3/70B_lora.yaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  apply_lora_to_mlp: False
+  apply_lora_to_output: False
+  lora_rank: 16
+  lora_alpha: 32
```
Comment (ebsmothers) [inline recipes/configs/llama3/70B_lora.yaml:19]: How are these defaults set? Are rank and alpha higher than our 7/8B defaults cause the embedding dim is larger?

File: recipes/configs/llama3/70B_lora.yaml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    model-00022-of-00030.safetensors,
+    model-00023-of-00030.safetensors,
+    model-00024-of-00030.safetensors,
+    model-00025-of-00030.safetensors,
+    model-00026-of-00030.safetensors,
+    model-00027-of-00030.safetensors,
+    model-00028-of-00030.safetensors,
+    model-00029-of-00030.safetensors,
+    model-00030-of-00030.safetensors,
+  ]
```
Comment (ebsmothers) [inline recipes/configs/llama3/70B_lora.yaml:59]: Lol we really need to have a way to generate these programmatically

**Review by ebsmothers — APPROVED** [review]

**Comment by musabgultekin** [comment]
@rohan-varma Have you tested if full-weight training works with 8x80GB ?

File: README.md (hunk 3 above)
Comment (rohan-varma) [reply README.md:118]: Yeah, maybe we can add it to the table?

File: recipes/configs/llama3/70B_lora.yaml (hunk 5 above)
Comment (rohan-varma) [reply recipes/configs/llama3/70B_lora.yaml:19]: I mostly copied these from https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama2/70B_lora.yaml. 

I'm not sure whether rank and alpha being higher is because of the embedding dim being larger - what do those have to do with each other?

**Comment by rohan-varma** [comment]
@musabgultekin I've dug into this a bit, and haven't gotten it into a full working state yet. This is something on our immediate list of priorities, so stay tuned!

**Comment by musab-mk** [comment]
Thank you! Will be following you!
